File: 2961.py
"""
* "모든 조합"을 구하는 문제입니다.
  * Why? 재료 (A, B, C)를 사용한 것과 재료 (A, C, B)를 사용한 것이 같은 결과를 가진다.
* 이때, 1개만 써도 되고, 2개만 써도 되고, ... N개를 써도 되니까
  * "모든 개수"에 대하여 "모든 조합"을 구하는 문제입니다.

[Example]
* A, B, C, D라는 재료가 있습니다.
* 길이 #1
  - A만 사용할 때
  - B만 사용할 때
  - C만 사용할 때
  - D만 사용할 때
* 길이 #2
  - (A, B)
  - (A, C)
  - (A, D)
  - (B, C)
  - (B, D)
  - (C, D)
* 길이 #3
  - (A, B, C)
  - (A, B, D)
  - (A, C, D)
  - (B, C, D)
* 길이 #4
  - (A, B, C, D)
"""

import sys
sys.setrecursionlimit(int(1e6))

n = int(input())
arr = []
for i in range(n):
    x, y = map(int, input().split())
    arr.append((x, y))

visited = [False] * n
result = []
answer = int(1e9)

def dfs(depth, start):
    global answer
    # 현재 조합에 대하여 결과 계산
    if depth >= 1:
        totalX = 1;
        totalY = 0;
        for i in result:
            x, y = arr[i]
            totalX *= x;
            totalY += y;
        answer = min(answer, abs(totalX - totalY))
    for i in range(start, n):
        if visited[i]: continue
        visited[i] = True
        result.append(i)
        dfs(depth + 1, i + 1)
        visited[i] = False
        result.pop()

dfs(0, 0)
print(answer)

File: 2961.java
import java.io.*;
import java.util.*;

public class Main {

  public static boolean[] visited = new boolean[10];
  public static Stack<Integer> result = new Stack<Integer>();
  public static int[][] arr = new int[10][2];
  public static int n;
  public static int answer = (int) 1e9;

  // N개의 재료 중에서 depth를 선택한 조합 계산
  public static void dfs(int depth, int start) {
    // "모든 개수"를 고려해야 하므로, depth가 1 이상이면 항상 처리
    // 현재 조합에 대한 처리
    if (depth >= 1) {
      int totalX = 1;
      int totalY = 0;
      for (int i: result) {
        int x = arr[i][0];
        int y = arr[i][01];
        totalX *= x;
        totalY += y;
      }
      answer = Math.min(answer, Math.abs(totalX - totalY));
    }
    for (int i = start; i < n; i++) {
      if (visited[i]) continue;
      // 방문 처리
      result.push(i);
      visited[i] = true;
      dfs(depth + 1, i + 1); // 재귀 함수 호출
      // 방문 처리 해제
      result.pop();
      visited[i] = false;
    }
  }
  
  public static void main(String[] args) throws IOException {
    Scanner scanner = new Scanner(System.in);
    n = scanner.nextInt();
    for (int i = 0; i < n; i++) {
      int x = scanner.nextInt();
      int y = scanner.nextInt();
      arr[i][0] = x;
      arr[i][1] = y;
    }
    dfs(0, 0);
    System.out.println(answer);
  }
}
